### PR TITLE
Turf fires

### PR DESCRIPTION
**Pull request by rye-rice**
## About The Pull Request

Port of https://github.com/hrzntal/horizon/pull/300
and https://github.com/hrzntal/horizon/pull/558
Sounds from https://github.com/tgstation/TerraGov-Marine-Corps 

With the extra feature of being able to color flames, as a proof of concept, pulse rifles drop flames
## Why It's Good For The Game

Admins can now blast vietnam music
This also makes flamethrowers less of a grief tool, as now they can work like a AOE denial weapon, and are easily countered by extingishers
image

## Changelog
:cl:Azarak
add: Turfs can now be ignited and will slowly burn, igniting anyone on them
add: Hotspots have 30% chance to ignite a turf, making fiery explosions leave cool residual flame.
balance: Flamethrowers have been reworked to use chemicals instead and now leave fire and hotspots on the turfs. They require an open reagent container instead of plasma tank now.
balance: Firealarms now trigger from 80C upwards (from 200C) or -10C downwards (from -3C). They also trigger immediately when they're on a turf fire
/:cl:

### CODE REVIEW

File: code/game/objects/effects/turf_fire.dm [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+			return
+		if(open_turf.air.return_temperature() < TURF_FIRE_REQUIRED_TEMP)
+			fire_power -= TURF_FIRE_POWER_LOSS_ON_LOW_TEMP
+		fire_power--
+		if(fire_power <= 0)
+			qdel(src)
+			return
+	open_turf.hotspot_expose(TURF_FIRE_TEMP_BASE + (TURF_FIRE_TEMP_INCREMENT_PER_POWER*fire_power), TURF_FIRE_VOLUME)
+	for(var/A in open_turf)
+		var/atom/AT = A
```
Comment (Kapu1178) [inline code/game/objects/effects/turf_fire.dm]:
```suggestion
	for(var/atom/movable/AT as anything in open_turf)
```

File: code/game/objects/items/flamethrower.dm [hunk 2]
```diff
@@ -241,17 +248,17 @@
...
+		QDEL_NULL(beaker)
 		return 1 //It hit the flamethrower, not them
 
-
-/obj/item/assembly/igniter/proc/flamethrower_process(turf/open/location)
-	location.hotspot_expose(700,2)
-
-/obj/item/assembly/igniter/proc/ignite_turf(obj/item/flamethrower/F,turf/open/location,release_amount = 0.05)
-	F.default_ignite(location,release_amount)
+/obj/item/flamethrower/proc/instant_refill()
```
Comment (Fikou) [inline code/game/objects/items/flamethrower.dm]: doesnt seem to be called anywhere

File: code/game/objects/effects/turf_fire.dm (hunk 1 above)
Comment (Fikou) [reply code/game/objects/effects/turf_fire.dm]: descript variable names would have been nice too

File: code/game/objects/effects/turf_fire.dm [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	anchored = TRUE
+	move_resist = INFINITY
+	light_range = 1.5
+	light_power = 1.5
+	light_color = LIGHT_COLOR_FIRE
+	mouse_opacity = FALSE
+	/// How much power have we got. This is treated like fuel, be it flamethrower liquid or any random thing you could come up with
+	var/fire_power = 20
+	/// Is it magical, if it is then it wont interact with atmos, and it will not loose power by itself. Mainly for adminbus events or mapping
+	var/magical = FALSE
```
Comment (Fikou) [inline code/game/objects/effects/turf_fire.dm]: i would suggest splitting this into atmos_interact and passive_loss (probably a better name?)

File: code/game/objects/effects/turf_fire.dm [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+		AT.fire_act(TURF_FIRE_TEMP_BASE + (TURF_FIRE_TEMP_INCREMENT_PER_POWER*fire_power), TURF_FIRE_VOLUME)
+	if(!magical)
+		if(prob(fire_power))
+			open_turf.burn_tile()
+		if(prob(6))
+			playsound(open_turf, 'sound/effects/comfyfire.ogg', 40, TRUE)
+		UpdateFireState()
+
+// /obj/effect/abstract/turf_fire/proc/on_entered(datum/source, atom/movable/AM) for crossed removal
+/obj/effect/abstract/turf_fire/Crossed(atom/movable/AM)
```
Comment (Fikou) [inline code/game/objects/effects/turf_fire.dm]: descript variable names

File: code/__DEFINES/is_helpers.dm [hunk 5]
```diff
@@ -25,6 +25,8 @@ GLOBAL_LIST_INIT(turfs_without_ground, typecacheof(list(
 
 #define isspaceturf(A) (istype(A, /turf/open/space))
 
+#define isopenspaceturf(A) (istype(A, /turf/open/openspace))
```
Comment (Fikou) [inline code/__DEFINES/is_helpers.dm:28]: this codebase doesnt seem to use openspace so not sure if this is necessary

File: code/modules/reagents/chemistry/holder.dm [hunk 6]
```diff
@@ -92,6 +92,13 @@
 		my_atom.reagents = null
 	my_atom = null
 
+/datum/reagents/proc/get_total_accelerant_quality()
+	var/quality = 0
+	for(var/i in reagent_list)
+		var/datum/reagent/reagent = i
```
Comment (Fikou) [inline code/modules/reagents/chemistry/holder.dm]:
```suggestion
	for(var/datum/reagent/reagent as anything in reagent_list)
```

File: code/modules/food_and_drinks/drinks/drinks/bottle.dm [hunk 7]
```diff
@@ -510,12 +508,11 @@
 	return
 
 /obj/item/reagent_containers/food/drinks/bottle/molotov/throw_impact(atom/hit_atom, datum/thrownthing/throwingdatum)
-	var/firestarter = 0
+	var/firestarter = FALSE
 	for(var/datum/reagent/R in reagents.reagent_list)
```
Comment (Fikou) [inline code/modules/food_and_drinks/drinks/drinks/bottle.dm]:
```suggestion
	for(var/datum/reagent/reagent as anything in reagents.reagent_list)
```

File: code/modules/food_and_drinks/drinks/drinks/bottle.dm [hunk 8]
```diff
@@ -510,12 +508,11 @@
...
-		for(var/A in accelerants)
-			if(istype(R,A))
-				firestarter = 1
-				break
+		if(R.accelerant_quality)
```
Comment (Fikou) [inline code/modules/food_and_drinks/drinks/drinks/bottle.dm]:
```suggestion
		if(reagent.accelerant_quality)
```

File: code/game/objects/effects/turf_fire.dm (hunk 4 above)
Comment (rye-rice) [reply code/game/objects/effects/turf_fire.dm]: Wouldn't it be better to keep it how it's named in the parent proc? Testing shows that it works when you change it to something more readable

File: code/game/objects/effects/turf_fire.dm (hunk 3 above)
Comment (rye-rice) [reply code/game/objects/effects/turf_fire.dm]: done

File: code/game/objects/items/flamethrower.dm (hunk 2 above)
Comment (rye-rice) [reply code/game/objects/items/flamethrower.dm]: removed

File: code/game/objects/effects/turf_fire.dm [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+		base_icon_state = "red"
+	else if(fire_color in allowed_colors)
+		base_icon_state = fire_color
+	else
+		hex_color = fire_color
+		color = fire_color
+		base_icon_state = "greyscale"
+
+	open_turf.turf_fire = src
+	SSturf_fire.fires[src] = TRUE
```
Comment (Fikou) [inline code/game/objects/effects/turf_fire.dm]: iirc associated lists are worse for performance, and you dont seem to be doing anything with the association, why not just do SSturf_fire.fires += src?

File: code/game/objects/effects/turf_fire.dm [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+			return
+	if(passive_loss)
+		if(open_turf.air.return_temperature() < TURF_FIRE_REQUIRED_TEMP)
+			fire_power -= TURF_FIRE_POWER_LOSS_ON_LOW_TEMP
+		fire_power--
+		if(fire_power <= 0)
+			qdel(src)
+			return
+	open_turf.hotspot_expose(TURF_FIRE_TEMP_BASE + (TURF_FIRE_TEMP_INCREMENT_PER_POWER*fire_power), TURF_FIRE_VOLUME)
+	for(var/atom/movable/AT as anything in open_turf)
```
Comment (Fikou) [inline code/game/objects/effects/turf_fire.dm]: descript var names

File: code/game/objects/items/flamethrower.dm [hunk 11]
```diff
@@ -184,40 +181,52 @@
...
+	power = my_fraction.get_total_accelerant_quality() * FLAMETHROWER_POWER_MULTIPLIER
+	if(power < REQUIRED_POWER_TO_FIRE_FLAMETHROWER)
+		audible_message("<span class='danger'>The [src] sputters.</span>")
+		playsound(src, 'sound/weapons/gun/flamethrower/flamethrower_empty.ogg', 50, TRUE, -3)
 		return
+	playsound(src, pick('sound/weapons/gun/flamethrower/flamethrower1.ogg','sound/weapons/gun/flamethrower/flamethrower2.ogg','sound/weapons/gun/flamethrower/flamethrower3.ogg' ), 50, TRUE, -3)
 	operating = TRUE
+	var/turfs_flamed = 0
 	var/turf/previousturf = get_turf(src)
 	for(var/turf/T in turflist)
```
Comment (Fikou) [inline code/game/objects/items/flamethrower.dm:204]: here too would be nice

File: code/modules/projectiles/projectile/beams.dm [hunk 12]
```diff
@@ -119,6 +119,10 @@
 			SSexplosions.medobj += target
 		else
 			SSexplosions.medturf += target
+	var/turf/targets_turf = target.loc
+	if(!(isopenturf(targets_turf)))
```
Comment (Fikou) [inline code/modules/projectiles/projectile/beams.dm]:
```suggestion
	if(!isopenturf(targets_turf))
```

File: code/game/objects/effects/turf_fire.dm (hunk 9 above)
Comment (rye-rice) [reply code/game/objects/effects/turf_fire.dm]: done

File: code/game/objects/items/flamethrower.dm (hunk 11 above)
Comment (rye-rice) [reply code/game/objects/items/flamethrower.dm:204]: turfs_being_flamed

File: code/game/objects/effects/turf_fire.dm (hunk 10 above)
Comment (rye-rice) [reply code/game/objects/effects/turf_fire.dm]: burning_atom

**Review by ZephyrTFA — APPROVED** [review]

**Pull request merged**